### PR TITLE
Layer dev

### PR DESCRIPTION
**Pull request by walrusmcd**
temp

### CODE REVIEW

**Comment by martinb35** [comment]
first!

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -425,9 +425,11 @@ if (WIN32)
 
     # set linker flags to minimize the binary size.
     if (MSVC)
-      foreach(type EXE SHARED)
-        set(CMAKE_${type}_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_${type}_LINKER_FLAGS_RELWITHDEBINFO} /OPT:REF,ICF,LBR")
-        set(CMAKE_${type}_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_${type}_LINKER_FLAGS_RELWITHDEBINFO} /INCREMENTAL:NO")
+      foreach(type EXE STATIC SHARED)
+        if (NOT type MATCHES STATIC)
```
Comment (martinb35) [inline cmake/CMakeLists.txt]: >STATIC [](start = 29, length = 6)

does this affect other binaries or just ours?

File: cmake/onnxruntime.cmake [hunk 2]
```diff
@@ -66,6 +66,7 @@ target_link_libraries(onnxruntime PRIVATE
     ${PROVIDERS_NUPHAR}
     ${PROVIDERS_DML}
     ${PROVIDERS_ACL}
+	${onnxruntime_winml}
```
Comment (martinb35) [inline cmake/onnxruntime.cmake]: >	 [](start = 0, length = 1)

spaces

File: cmake/winml.cmake [hunk 3]
```diff
@@ -196,6 +200,10 @@ add_dependencies(winml_lib_core winml_api_native_internal)
 target_link_libraries(winml_lib_core PRIVATE wil)
 target_link_libraries(winml_lib_core PRIVATE onnxruntime_providers_dml)
 
+# add it to the onnxruntime shared library
+set(onnxruntime_winml windowsapp.lib -WHOLEARCHIVE:$<TARGET_FILE:winml_lib_core>)
+list(APPEND onnxruntime_EXTERNAL_DEPENDENCIES winml_lib_core)
+
```
Comment (martinb35) [inline cmake/winml.cmake]: is this going to live here in the final version? seems odd to add things to a different project

File: cmake/winml.cmake [hunk 4]
```diff
@@ -462,26 +471,26 @@ if("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
...
+#target_link_libraries(winml_dll PRIVATE onnxruntime_common)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_graph)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_framework)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_mlas)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_optimizer)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_providers)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_providers_dml)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_session)
+#target_link_libraries(winml_dll PRIVATE onnxruntime_util)
+#target_link_libraries(winml_dll PRIVATE onnx_proto)
```
Comment (martinb35) [inline cmake/winml.cmake]: why not just delete these instead of commenting?

File: cmake/winml.cmake [hunk 5]
```diff
@@ -462,26 +471,26 @@ if("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
...
+target_link_libraries(winml_dll PRIVATE onnxruntime)
 target_link_libraries(winml_dll PRIVATE re2)
 target_link_libraries(winml_dll PRIVATE wil)
+target_link_libraries(winml_dll PRIVATE windowsapp.lib)
 target_link_libraries(winml_dll PRIVATE winml_lib_api)
-target_link_libraries(winml_dll PRIVATE winml_lib_core)
+#target_link_libraries(winml_dll PRIVATE winml_lib_core)
```
Comment (martinb35) [inline cmake/winml.cmake]: delete

File: cmake/winml.cmake [hunk 6]
```diff
@@ -462,26 +471,26 @@ if("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
...
 target_link_libraries(winml_dll PRIVATE winml_lib_image)
 target_link_libraries(winml_dll PRIVATE winml_lib_telemetry)
-target_link_libraries(winml_dll PRIVATE onecoreuap_apiset.lib)
```
Comment (martinb35) [inline cmake/winml.cmake]: >onecoreuap_apiset [](start = 40, length = 17)

intentional? you replaced onecoreuap_apiset.lib with windowsapp.lib

File: include/onnxruntime/core/common/status.h [hunk 7]
```diff
@@ -74,6 +77,39 @@ inline const char* StatusCodeToString(StatusCode status) noexcept {
...
+    case StatusCode::NOT_IMPLEMENTED:
+        return E_NOTIMPL;
+    case StatusCode::INVALID_GRAPH:
+        return __HRESULT_FROM_WIN32(ERROR_FILE_CORRUPT);
+    case StatusCode::EP_FAIL:
+        return __HRESULT_FROM_WIN32(ERROR_INTERNAL_ERROR);
+    default:
+        return E_FAIL;
+    }
+}
```
Comment (martinb35) [inline include/onnxruntime/core/common/status.h]: this exists somewhere else, too right? in winml somewhere. can we share it instead of copying it?

Comment (martinb35) [reply include/onnxruntime/core/common/status.h]: nm. I see now that you removed it from the winml spot

---
In reply to: [347040246](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347040246) [](ancestors = 347040246)

File: winml/api/Windows.AI.MachineLearning.idl [hunk 8]
```diff
@@ -334,6 +334,18 @@ namespace Windows.AI.MachineLearning
...
+        {
+            //! allows creation of feature descriptors outside the runtime (immutable)
+            [method_name("Create")] MapFeatureDescriptor(
+                String Name,
+                String Description,
+                Boolean IsRequired,
+                TensorKind KeyKind,
+                ILearningModelFeatureDescriptor ValueDescriptor
+            );
+        }
```
Comment (martinb35) [inline winml/api/Windows.AI.MachineLearning.idl]: do we need to do API review?

Comment (martinb35) [reply winml/api/Windows.AI.MachineLearning.idl]: also can you explain what these new methods are for?

---
In reply to: [347040835](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347040835) [](ancestors = 347040835)

File: winml/lib/Api/LearningModel.h [hunk 9]
```diff
@@ -120,13 +116,10 @@ struct LearningModel : LearningModelT<LearningModel> {
       winml::ILearningModelFeatureDescriptor descriptor,
       bool& use_fp16);
 
-  void
-  OverrideShapeInferenceMethods();
```
Comment (martinb35) [inline winml/lib/Api/LearningModel.h]: >OverrideShapeInferenceMethods [](start = 2, length = 29)

where did this method go?

File: winml/lib/Api/LearningModelBinding.cpp [hunk 10]
```diff
@@ -215,98 +218,103 @@ void LearningModelBinding::Split(
   throw hresult_not_implemented();
 }
 
-onnxruntime::IOBinding& LearningModelBinding::BindingCollection() {
-  return *m_lotusBinding;
+_winmla::IIOBinding* LearningModelBinding::BindingCollection() {
+    _winmla::IIOBinding* p;
+    m_lotusBinding.copy_to(&p);
+    return p;
```
Comment (martinb35) [inline winml/lib/Api/LearningModelBinding.cpp]: is this a copy or an add-ref?

File: winml/lib/Api/LearningModelSession.cpp [hunk 11]
```diff
@@ -303,13 +200,14 @@ LearningModelSession::EvaluateFeaturesAsync(
...
-static onnxruntime::IOBinding&
+static _winmla::IIOBinding*
 GetIOBinding(
     winrt::com_ptr<winmlp::LearningModelBinding> binding_impl,
     winml::LearningModel& model) {
   // Get the IOBinding Collection, and bound outputs
-  auto& io_binding = binding_impl->BindingCollection();
-  auto& bound_output_names = io_binding.GetOutputNames();
+    com_ptr<_winmla::IIOBinding> io_binding;
+    io_binding.attach(binding_impl->BindingCollection());
```
Comment (martinb35) [inline winml/lib/Api/LearningModelSession.cpp]: spacing

File: winml/lib/Api/LearningModelSession.cpp [hunk 12]
```diff
@@ -362,15 +259,16 @@ LearningModelSession::Run(
   // TODO : set the run_options
   onnxruntime::RunOptions run_options;
 
-  auto& io_binding = GetIOBinding(binding_impl, model_);
+  com_ptr<_winmla::IIOBinding> io_binding;
+  io_binding.attach(GetIOBinding(binding_impl, model_));
 
   // Invoke run on the ORT session.
-  WINML_THROW_IF_NOT_OK(inference_session_->Run(run_options, io_binding));
+  WINML_THROW_IF_FAILED(inference_session_->Run(&run_options, io_binding.get()));
```
Comment (martinb35) [inline winml/lib/Api/LearningModelSession.cpp]: >WINML_THROW_IF_FAILED [](start = 2, length = 21)

NOT_OK is not the same as FAILED - why this change?

File: winml/lib/Api.Core/WinMLAdapter.cpp [hunk 13]
```diff
@@ -0,0 +1,840 @@
...
+  Load(std::unique_ptr<ONNX_NAMESPACE::ModelProto> p_model_proto) {
+    return onnxruntime::InferenceSession::Load(std::move(p_model_proto));
+  }
+};
+
+
+
+// class AbiSafeTensor
+//
+class AbiSafeTensor : public Microsoft::WRL::RuntimeClass <
```
Comment (martinb35) [inline winml/lib/Api.Core/WinMLAdapter.cpp]: >AbiSafe [](start = 6, length = 7)

are these names temporary? I don't love the "ABISafe" prefix...

File: winml/lib/Api.Core/WinMLAdapter.cpp [hunk 14]
```diff
@@ -0,0 +1,840 @@
...
+    Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::ClassicCom>,
+    ITensor> {
+private:
+    onnxruntime::Tensor& tensor_;   // weak ref
+    ComPtr<IOrtValue> value_;       // strong ref
```
Comment (martinb35) [inline winml/lib/Api.Core/WinMLAdapter.cpp]: I believe our code convention puts data members at the end of a class, right?

File: winml/lib/Api.Core/WinMLAdapter.cpp [hunk 15]
```diff
@@ -0,0 +1,840 @@
...
+            &file_descriptor,
+            path,
+            O_RDONLY | _O_SEQUENTIAL | _O_BINARY,
+            _SH_DENYWR,
+            _S_IREAD | _S_IWRITE);
+
+        THROW_HR_IF_MSG(
+            E_FAIL,
+            0 > file_descriptor,
+            "Failed");  //errno
```
Comment (martinb35) [inline winml/lib/Api.Core/WinMLAdapter.cpp]: can you integrate the file not found change I just made?

File: winml/lib/Api.Core/WinMLAdapter.cpp [hunk 16]
```diff
@@ -0,0 +1,840 @@
...
+    }
+
+
+};
+
+extern "C"
+HRESULT STDMETHODCALLTYPE OrtGetWinMLAdapter(IWinMLAdapter** adapter) {
+    // make an adapter instance
+    Microsoft::WRL::ComPtr<WinMLAdapter> adapterptr = wil::MakeOrThrow<WinMLAdapter>();
+    return adapterptr.CopyTo(__uuidof(IWinMLAdapter), (void **)adapter);
```
Comment (martinb35) [inline winml/lib/Api.Core/WinMLAdapter.cpp]: >(void **) [](start = 54, length = 9)

static_cast?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (walrusmcd) [reply cmake/CMakeLists.txt]: adding the STATIC in here makes it affect all static link libs.    it seemed to make sense to apply everywhere.    the problem is the linker has to take 2 passes if it needs to use LTCG and you didn't tell it before hand.  so all the static link libs where starting their link phase, then resetting once they realized they needed LTCG.   

---
In reply to: [347038839](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347038839) [](ancestors = 347038839)

File: cmake/onnxruntime.cmake (hunk 2 above)
Comment (walrusmcd) [reply cmake/onnxruntime.cmake]: good catch !

---
In reply to: [347039014](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347039014) [](ancestors = 347039014)

File: cmake/winml.cmake (hunk 6 above)
Comment (walrusmcd) [reply cmake/winml.cmake]: yeah, not sure why i did that.   

---
In reply to: [347039801](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347039801) [](ancestors = 347039801)

File: winml/lib/Api.Core/WinMLAdapter.cpp (hunk 16 above)
Comment (walrusmcd) [reply winml/lib/Api.Core/WinMLAdapter.cpp]: static_cast doesn't work to void ** .  fun, right?

[https://stackoverflow.com/questions/3625410/c-static-cast-from-float-to-void](https://stackoverflow.com/questions/3625410/c-static-cast-from-float-to-void)


---
In reply to: [347044742](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347044742) [](ancestors = 347044742)

File: winml/lib/Api/LearningModelSession.cpp (hunk 12 above)
Comment (walrusmcd) [reply winml/lib/Api/LearningModelSession.cpp]: nope.   not the same.  NOT_OK is for status.IsOK() checks.    This moved from an InferenceSession (ort type) to a IInferenceSession (com_ptr) and uses HRESULTS now across the ABI.    So it had to change from NOT_OK to FAILED.    

---
In reply to: [347042739](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347042739) [](ancestors = 347042739)

File: winml/lib/Api.Core/WinMLAdapter.cpp (hunk 13 above)
Comment (walrusmcd) [reply winml/lib/Api.Core/WinMLAdapter.cpp]: temporary.

---
In reply to: [347043931](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347043931) [](ancestors = 347043931)

File: winml/lib/Api.Core/WinMLAdapter.cpp (hunk 14 above)
Comment (walrusmcd) [reply winml/lib/Api.Core/WinMLAdapter.cpp]: probably.    i changed it to the top when the classes are coded inside the CPP file (also not something we normally do) .  when you put them at the bottom they are much harder to find.

---
In reply to: [347044267](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347044267) [](ancestors = 347044267)

File: winml/lib/Api.Core/WinMLAdapter.cpp (hunk 15 above)
Comment (walrusmcd) [reply winml/lib/Api.Core/WinMLAdapter.cpp]: how do i do that?  

---
In reply to: [347044532](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347044532) [](ancestors = 347044532)

File: winml/lib/Api/LearningModel.h (hunk 9 above)
Comment (walrusmcd) [reply winml/lib/Api/LearningModel.h]: IWinMLAdapter::OverrideSchemaInferenceFunctions()

---
In reply to: [347041742](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347041742) [](ancestors = 347041742)

File: winml/lib/Api/LearningModelBinding.cpp (hunk 10 above)
Comment (walrusmcd) [reply winml/lib/Api/LearningModelBinding.cpp]: copy_to bumps the ref, so it is returning a strong ref.  the caller needs to handle it.    in this case there is one caller, in learningmodelsession.cpp line #206

---
In reply to: [347042063](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347042063) [](ancestors = 347042063)

File: winml/api/Windows.AI.MachineLearning.idl (hunk 8 above)
Comment (walrusmcd) [reply winml/api/Windows.AI.MachineLearning.idl]: yep.  we will need an API review.  these new methogs allow creation of feature descriptors outside the runtime (the code comment there).   basically before this change you could never create these on their own but had to have the runtime make then inside of models and return them.   this change adds a factory so that other people can create them.   ORT creates them in this PR.  and in the future people like Vision Skills could create then also. it makes them reusable currencies.

---
In reply to: [347040969](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347040969) [](ancestors = 347040969,347040835)

File: winml/lib/Api/LearningModelSession.cpp (hunk 11 above)
Comment (walrusmcd) [reply winml/lib/Api/LearningModelSession.cpp]: good catch !

---
In reply to: [347042505](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347042505) [](ancestors = 347042505)

File: winml/lib/Api.Core/WinMLAdapter.cpp (hunk 15 above)
Comment (walrusmcd) [reply winml/lib/Api.Core/WinMLAdapter.cpp]: i figured it out.   the change is here: [https://dev.azure.com/microsoft/WindowsAI/_git/WindowsAI/pullrequest/4014934?_a=files&path=%2Fwinml%2Flib%2FCore%2FModelInfo.cpp](https://dev.azure.com/microsoft/WindowsAI/_git/WindowsAI/pullrequest/4014934?_a=files&path=%2Fwinml%2Flib%2FCore%2FModelInfo.cpp)

---
In reply to: [347543900](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347543900) [](ancestors = 347543900,347044532)

File: cmake/winml.cmake (hunk 3 above)
Comment (walrusmcd) [reply cmake/winml.cmake]: that was the idea.    onnxruntime_EXTERNAL_DEPENDENCIES is a property setup just for this purpose.  for other cmake files to add things to the the final shared_lib

---
In reply to: [347039511](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347039511) [](ancestors = 347039511)

File: cmake/winml.cmake (hunk 6 above)
Comment (martinb35) [reply cmake/winml.cmake]: this doesn't seem resolved yet...

---
In reply to: [347539379](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347539379) [](ancestors = 347539379,347039801)

File: winml/lib/Api.Core/WinMLAdapter.cpp (hunk 16 above)
Comment (martinb35) [reply winml/lib/Api.Core/WinMLAdapter.cpp]: reinterpret! that's what I meant. my bad.

---
In reply to: [347542211](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347542211) [](ancestors = 347542211,347044742)

File: cmake/onnxruntime.cmake [hunk 17]
```diff
@@ -66,6 +66,7 @@ target_link_libraries(onnxruntime PRIVATE
     ${PROVIDERS_NUPHAR}
     ${PROVIDERS_DML}
     ${PROVIDERS_ACL}
+    ${onnxruntime_winml}
```
Comment (ryanlai2) [inline cmake/onnxruntime.cmake]: >onnxruntime_winml [](start = 6, length = 17)

Should the naming have adapter / core in there? I had to check in winml.cmake to double check which lib(s) were getting included

File: winml/lib/Api.Core/WinMLAdapter.cpp [hunk 18]
```diff
@@ -0,0 +1,851 @@
...
+        }
+        return S_OK;
+    }
+
+    const std::vector<std::string>& STDMETHODCALLTYPE GetOutputNames() override {
+        return binding_->GetOutputNames();
+    }
+    std::vector<IOrtValue*>& STDMETHODCALLTYPE GetOutputs() override {
+        auto output_inner = binding_->GetOutputs();
+        outputs_.clear();
```
Comment (ryanlai2) [inline winml/lib/Api.Core/WinMLAdapter.cpp]: Also need to clear outputs_weak vector holding a vector of weak ptrs :)

File: winml/lib/Api.Core/WinMLAdapter.cpp [hunk 19]
```diff
@@ -0,0 +1,865 @@
...
+#include "core/providers/dml/DmlExecutionProvider/inc/DmlExecutionProvider.h"
+#include "core/providers/dml/GraphTransformers/GraphTransformerHelpers.h"
+#include "core/providers/dml/OperatorAuthorHelper/SchemaInferenceOverrider.h"
+
+#include "LearningModelDevice.h"
+#include "TensorFeatureDescriptor.h"
+#include "ImageFeatureDescriptor.h"
+#include "api.image/inc/D3DDeviceCache.h"
+
+#include "PheonixSingleton.h"
```
Comment (ryanlai2) [inline winml/lib/Api.Core/WinMLAdapter.cpp]: Already included "PheonixSingleton.h" up above
